### DISCUSSION TITLE
Is possible change colors on-the-fly without compile?

### DISCUSSION THREAD

**Discussion by otavioserra** [fomantic/Fomantic-UI#2123] [Behaviors (Api, Form Validation, Visibility)]
I want to create some kind of color picker to my users change colors in the application and apply it to Fomantic UI (2.8.8). I can't found any kind of examples that do that in the documentation or here. I only found forms to change themes colors when it's on developer ambient. So, to change colors I need to compile CSS always. This is good to prepare application, but to change colors dynamic isn't. It was my mistake or there are some form to do that?

Thanks

**Comment by lubber-de** [comment]
If you want to create FUI style classnames for new colors (mintgreen, moonblue, etc.) you have to recompile, because the colors are build into each element out of the central [colors.less](https://github.com/fomantic/Fomantic-UI/blob/develop/src/themes/default/globals/colors.less) file
I explained that a bit more at https://github.com/fomantic/Fomantic-UI/issues/402#issuecomment-454913211

The FUI approach is to declare a color name which affects all elements. But as many elements have different shades of colors in parts of their appearance this is not a single color to be taken care of.

What exactly is your use case to change colors of each element on the fly? If it's only for a single element, you might try to use css variables only for that.

**Reply by otavioserra** [reply]
@lubber-de I think css variables is the answer because my users can change colors of some elements and I'll need create color picker on-the-fly. But I tried to change semantic/src/site/globals/site.variables and put some css variables but it throwed errors on gulp build.

**CSS Variable: --fui-primary-color**
```
@primaryColor        : var(--fui-primary-color);
```
**semantic/src/site/globals/site.variables**
```
/*******************************
    User Global Variables
*******************************/

/*---  Primary and Secondary Colors  ---*/
@primaryColor        : var(--fui-primary-color);
@secondaryColor      : var(--fui-secondary-color);

@lightPrimaryColor   : @lightBlue;
@lightSecondaryColor : @lightBlack;
```

**command line:**

```
C:\nodeProjects\fomanticUI\node_modules\fomantic-ui> npx gulp build
[16:10:12] Using gulpfile C:\nodeProjects\fomanticUI\node_modules\fomantic-ui\gulpfile.js
[16:10:12] Starting 'build'...
Building Semantic
[16:10:12] Starting 'build-css'...
[16:10:12] Starting 'Building uncompressed CSS'...
[16:10:12] Starting 'Building compressed CSS'...
(node:7424) UnhandledPromiseRejectionWarning: Error: Error in plugin "gulp-less"
Message:
    Error evaluating function `darken`: Argument cannot be evaluated to a color in file C:\nodeProjects\fomanticUI\semantic\src\themes\default\globals\site.variables line no. 372
Details:
    type: Runtime
    filename: C:\nodeProjects\fomanticUI\semantic\src\themes\default\globals\site.variables
    index: 10400
    line: 372
    column: 22
    callLine: NaN
    callExtract: undefined
    extract: /*--- Shadows ---*/,@primaryRibbonShadow: darken(@primaryColor, 10);,@secondaryRibbonShadow: darken(@secondaryColor, 10);
    lineNumber: 372
    fileName: C:\nodeProjects\fomanticUI\semantic\src\themes\default\globals\site.variables

    at Transform.errorHandler (C:\nodeProjects\fomanticUI\node_modules\fomantic-ui\tasks\config\tasks.js:125:19)
    at Transform.emit (events.js:194:15)
    at Transform.EventEmitter.emit (domain.js:441:20)
    at errorOrDestroy (C:\nodeProjects\fomanticUI\node_modules\gulp-less\node_modules\readable-stream\lib\internal\streams\destroy.js:98:101)
    at onwriteError (C:\nodeProjects\fomanticUI\node_modules\gulp-less\node_modules\readable-stream\lib\_stream_writable.js:430:5)
    at onwrite (C:\nodeProjects\fomanticUI\node_modules\gulp-less\node_modules\readable-stream\lib\_stream_writable.js:450:11)
    at WritableState.onwrite (C:\nodeProjects\fomanticUI\node_modules\gulp-less\node_modules\readable-stream\lib\_stream_writable.js:160:5)
    at Transform.afterTransform (C:\nodeProjects\fomanticUI\node_modules\gulp-less\node_modules\readable-stream\lib\_stream_transform.js:89:3)
    at C:\nodeProjects\fomanticUI\node_modules\gulp-less\index.js:109:14
(node:7424) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:7424) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:7424) UnhandledPromiseRejectionWarning: Error: Error in plugin "gulp-less"
Message:
    Error evaluating function `darken`: Argument cannot be evaluated to a color in file C:\nodeProjects\fomanticUI\semantic\src\themes\default\globals\site.variables line no. 372
Details:
    type: Runtime
    filename: C:\nodeProjects\fomanticUI\semantic\src\themes\default\globals\site.variables
    index: 10400
    line: 372
    column: 22
    callLine: NaN
    callExtract: undefined
    extract: /*--- Shadows ---*/,@primaryRibbonShadow: darken(@primaryColor, 10);,@secondaryRibbonShadow: darken(@secondaryColor, 10);
    lineNumber: 372
    fileName: C:\nodeProjects\fomanticUI\semantic\src\themes\default\globals\site.variables

    at Transform.errorHandler (C:\nodeProjects\fomanticUI\node_modules\fomantic-ui\tasks\config\tasks.js:125:19)
    at Transform.emit (events.js:194:15)
    at Transform.EventEmitter.emit (domain.js:441:20)
    at errorOrDestroy (C:\nodeProjects\fomanticUI\node_modules\gulp-less\node_modules\readable-stream\lib\internal\streams\destroy.js:98:101)
    at onwriteError (C:\nodeProjects\fomanticUI\node_modules\gulp-less\node_modules\readable-stream\lib\_stream_writable.js:430:5)
    at onwrite (C:\nodeProjects\fomanticUI\node_modules\gulp-less\node_modules\readable-stream\lib\_stream_writable.js:450:11)
    at WritableState.onwrite (C:\nodeProjects\fomanticUI\node_modules\gulp-less\node_modules\readable-stream\lib\_stream_writable.js:160:5)
    at Transform.afterTransform (C:\nodeProjects\fomanticUI\node_modules\gulp-less\node_modules\readable-stream\lib\_stream_transform.js:89:3)
    at C:\nodeProjects\fomanticUI\node_modules\gulp-less\index.js:109:14
(node:7424) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
[16:10:13] The following tasks did not complete: build, build-css, <series>, <series>, Building uncompressed CSS, Building compressed CSS
[16:10:13] Did you forget to signal async completion?
PS C:\nodeProjects\fomanticUI\node_modules\fomantic-ui>
```

**Reply by otavioserra** [reply]
@lubber-de I can refer css variables in other way? Or I made some mistake?

**Reply by lubber-de** [reply]
Here is the problem
```
@primaryRibbonShadow: darken(@primaryColor, 10);
```
LESS has color manipulating functions which only work with absolute colors like RGB values.
But as you declared `@primaryColor` as a CSS variable the LESS functions break.

Of course you can use CSS variables as LESS variables like you did for primaryColor, BUT you would also have to adjust everything else in your theme, which refers to those variables and adjust them as well.

In the above mentioned example you would have to change `@primaryRobboinShadow` to another CSS Variable. Beware: There are LOTS of variables in the default theme which make use of the LESS functions and would have to get CSS variables instead.

**Reply by deanp70** [reply]
Hey @lubber-de, I hope it's ok I'm barging into this discussion, but I found it trying to search for an issue I have building Fomantic, and maybe I can get your help and if someone else searches for this, they'll find the solution here too.

I'm trying to build Fomantic (without any changes), and I get this error as well. I'm not sure how that makes sense, since I'm assuming to create the distribution, you should just follow the instructions. Specifically the error I get is:
```
(node:80316) UnhandledPromiseRejectionWarning: Error: Error in plugin "gulp-less"
Message:
    Error evaluating function `darken`: Argument cannot be evaluated to a color in file {path}/src/themes/default/modules/popup.variables line no. 53
```

Any ideas on how I can resolve this? I can't find anything searching through the interwebz
Thanks in advance 🙏

**Reply by lubber-de** [reply]
Mmh, if you really did not change the mentioned file, and freshly installed from nom then i believe you have an old LESS version by whatever reason.
How did you install fomantic? Via bpl Install or did you clone the repository? However, try to go into the fomantic  folder and try `npm install` which should take care of the proper less version needed by fui (3.7+ or 4.x if you use fui 2.8.8)
After that try `npx gulp build` to build fui which should give no errors anymore.

If you happen to just use semantic.less in your less project, then this should be same reason (very old less version) in this case update your less compiler version as well

**Reply by deanp70** [reply]
Thanks for the quick response @lubber-de! Would your answer change if I added the bit of information that I'm working with an M1 laptop. I don't think it should, but maybe there's some known issue.

One of the strange things that I couldn't resolve is after `npm install` it seemed that many of the packages were not installed and I had to manually add them in order to build the package.

Will try again with a clean install

Just to be sure:
after initializing an empty project I should `npm install fomantic-ui` then after it's done it will go through the install, I should choose default all, then go to node modules and do `npx gulp build`?

Am I missing something?

**Reply by lubber-de** [reply]
It does not matter what kind of laptop you use.
Yes, basically `npm install --ignore-scripts fomantic-ui` and `npx gulp build` should do it .

**Reply by deanp70** [reply]
It indeed works now. Thx, I probably screwed something up along the way. Thank you very much for your help and responsiveness!

**Comment by lubber-de** [comment]
If it's only for colors then try to use CSS Variables directly in the central color file 
https://github.com/fomantic/Fomantic-UI/blob/develop/src/themes/default/globals/colors.less

**Comment by otavioserra** [comment]
@lubber-de 

I will try to explain my use case in more detail: I have a platform that any user, via the configuration panel, can choose any color using a color picker. With any color chosen, I would like to apply the same color to a FomanticUI element dynamically. For example, the user chooses a specific color #456389 and that color is dynamically applied to the primary button (#456389 - this color is purely hypothetical).

When you said to use CSS variables, I could see in the CSS documentation that I can define a variable in one CSS file and then in another CSS file get the value of this variable. So, when my user makes a choice of color, I create a variable.css file with the definition of this value. So I create the following 'variables.css' file with this reference:

```
:root {
  --fui-primary-color: #456389;
}
```
Then I need to reference this same color in 'semantic.css' somehow so that this color can be dynamically read by FomanticUI. In the previous case, I tried to do this using site.variables . But for some reason, LESS does not support reading CSS variables in the format "var(--variable-name)" as per the CSS documentation like this case I tried to compile:
```
@primaryColor        : var(--fui-primary-color);
```
So, is there any way to reference these dynamic colors in FomanticUI without having to compile? In other words, modify colors on-the-fly without having to define predefined color palettes?

**Reply by lubber-de** [reply]
Please see my [answer above](https://github.com/fomantic/Fomantic-UI/discussions/2123#discussioncomment-1444490) where i explained that LESS functions do not work with css variables. If you want css variables make sure that every usage inside of fomantic uses css variables instead of calculating them at build time.
